src/swap.ts

```diff
@@ -0,0 +1,291 @@
+import { purry } from './purry';
+import { isString } from './isString';
+import { isArray } from './isArray';
+import { IterableContainer } from './_types';
+
+type isEqual<A, B> = A extends B ? (B extends A ? true : false) : false;
+
+type isLessThan<A extends number, B extends number> = isEqual<A, B> extends true
+  ? false
+  : 0 extends A
+  ? true
+  : 0 extends B
+  ? false
+  : isLessThan<Subtract<A, 1>, Subtract<B, 1>>;
+
+type BuildTuple<
+  L extends number,
+  T extends ReadonlyArray<unknown> = []
+> = T['length'] extends L ? T : BuildTuple<L, [...T, unknown]>;
+
+type Subtract<A extends number, B extends number> = BuildTuple<A> extends [
+  ...infer U,
+  ...BuildTuple<B>
+]
+  ? U['length']
+  : never;
+
+type IsNonNegativeInteger<T extends number> = number extends T
+  ? false
+  : `${T}` extends `-${string}` | `${string}.${string}`
```

Comment:
I know that for completeness the fractional part makes sense, but we don't actually need it here because it's not an input we need to care about. Nobody would try to swap item and index 9.75 and if they do I don't care about the type.
```suggestion
  : `${T}` extends `-${string}`
```